src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.vb

```diff
@@ -1339,20 +1343,50 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-                Return False
+        ''' <summary>
+        ''' Is this a floating-point operation that results in a whole number, rendering a following rounding operation redundant?
+        ''' </summary>
+        Private Function ReturnsWholeNumberDouble(node As BoundCall) As Boolean
+            Dim methodName As String = node.Method.Name
+            If "Ceiling".Equals(methodName) Then
+                Return node.Method = Me.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Math__CeilingDouble)
+            ElseIf "Floor".Equals(methodName) Then
+                Return node.Method = Me.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Math__FloorDouble)
+            ElseIf "Round".Equals(methodName) Then
+                Return node.Method = Me.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Math__RoundDouble)
+            ElseIf "Int".Equals(methodName) Then
+                Select Case node.Type.SpecialType
+                    Case SpecialType.System_Single
+                        Return node.Method = Me.Compilation.GetWellKnownTypeMember(WellKnownMember.Microsoft_VisualBasic_Conversion__IntSingle)
+                    Case SpecialType.System_Double
+                        Return node.Method = Me.Compilation.GetWellKnownTypeMember(WellKnownMember.Microsoft_VisualBasic_Conversion__IntDouble)
+                    Case Else
+                        Return False
+                End Select
             End If
 
-            Select Case node.Type.SpecialType
-                Case SpecialType.System_Single
-                    Return node.Method = Me.Compilation.GetWellKnownTypeMember(WellKnownMember.Microsoft_VisualBasic_Conversion__FixSingle)
-                Case SpecialType.System_Double
-                    Return node.Method = Me.Compilation.GetWellKnownTypeMember(WellKnownMember.Microsoft_VisualBasic_Conversion__FixDouble)
-                Case Else
-                    Return False
-            End Select
+            Return False
+        End Function
+
+        ''' <summary>
+        ''' Is this a floating-point truncation operation that would be redundant if followed by a truncation to an integral type?
+        ''' </summary>
+        Private Function IsFloatingTruncation(node As BoundCall) As Boolean
+            Dim methodName As String = node.Method.Name
+            If "Fix".Equals(methodName) Then
+                Select Case node.Type.SpecialType
+                    Case SpecialType.System_Single
+                        Return node.Method = Me.Compilation.GetWellKnownTypeMember(WellKnownMember.Microsoft_VisualBasic_Conversion__FixSingle)
+                    Case SpecialType.System_Double
+                        Return node.Method = Me.Compilation.GetWellKnownTypeMember(WellKnownMember.Microsoft_VisualBasic_Conversion__FixDouble)
+                    Case Else
+                        Return False
+                End Select
+            ElseIf "Truncate".Equals(methodName) Then
```

Comment:
Would it be worth it to also add a check on special type (before checking the method), as a shortcut?

Reply:
There is no way to go from a type to its special type. There is only such a thing for well-known types. The best we could do is look up the special type and see if the type is the same. That is as expensive as the test we already do (once the cache is filled).

---
In reply to: [190118936](https://github.com/dotnet/roslyn/pull/26948#discussion_r190118936) [](ancestors = 190118936)